### PR TITLE
update accessibility mode on window reload

### LINKED ISSUE #167583: NVDA is not recognized on Windows After Reload Of VS Code
Testing #167159

Steps:
* Start NVDA
* Start VS Code Insiders
* Screen reader is detected.
* Reload VS Code
* Screen reader is not detected anymore.

### PR DESCRIPTION
**Pull request by meganrogge**
fixes https://github.com/microsoft/vscode/issues/167583

### CODE REVIEW

**Comment by meganrogge** [comment]
Oh ok great then this should be good to review and merge

File: src/vs/code/electron-main/app.ts [hunk 1]
```diff
@@ -1137,6 +1139,8 @@ export class CodeApplication extends Disposable {
 
 		// Crash reporter
 		this.updateCrashReporterEnablement();
+
+
```
Comment (meganrogge) [inline src/vs/code/electron-main/app.ts:1143]:
```suggestion
```

File: src/vs/code/electron-main/app.ts [hunk 2]
```diff
@@ -944,7 +947,6 @@ export class CodeApplication extends Disposable {
 				return false;
 			}
 		});
```
Comment (meganrogge) [inline src/vs/code/electron-main/app.ts:949]:
```suggestion
		});

```

**Review by bpasero — CHANGES_REQUESTED** [review]
Why does this have to be a IPC message sent from `app.ts` and why can this not be one of the configuration options that we pass into a window here:

https://github.com/microsoft/vscode/blob/1c85aeab6777c0cd07d847a8c7fa0daff22473be/src/vs/platform/windows/electron-main/windowsMainService.ts#L1329

**Review by Adisa101 — APPROVED** [review]

**Comment by bpasero** [comment]
@meganrogge `windowImpl` is still in `electron-main` and has access to Electron `app` module, see for example: 

https://github.com/microsoft/vscode/blob/b92894493e6f971a3c71912d823e699d3a05d643/src/vs/platform/windows/electron-main/windowImpl.ts#L485

Same goes for `windowsMainService` if we want to put this into the config.

Bottom line: instead of sending an IPC message to the window, I would send this information to the window as part of the config when opening.

**Comment by meganrogge** [comment]
it turns out, we're already passing that in as an option, but not updating the status bar item on reload - so the latter is actually what's needed 

https://github.com/microsoft/vscode/blob/a3dd458dc0e92eccee2390bba81615386447a8d3/src/vs/platform/windows/electron-main/windowsMainService.ts#L1380

File: src/vs/code/electron-main/app.ts [hunk 3]
```diff
@@ -760,6 +765,7 @@ export class CodeApplication extends Disposable {
 
 		// Native host (main & shared process)
 		this.nativeHostMainService = accessor.get(INativeHostMainService);
+
```
Comment (meganrogge) [inline src/vs/code/electron-main/app.ts:768]:
```suggestion
```

File: src/vs/workbench/browser/parts/editor/accessibilityStatus.ts [hunk 4]
```diff
@@ -25,10 +25,9 @@ export class AccessibilityStatus extends Disposable implements IWorkbenchContrib
 		@IConfigurationService private readonly configurationService: IConfigurationService,
 		@INotificationService private readonly notificationService: INotificationService,
 		@IAccessibilityService private readonly _accessibilityService: IAccessibilityService,
-		@IStatusbarService private readonly statusbarService: IStatusbarService
+		@IStatusbarService private readonly statusbarService: IStatusbarService,
```
Comment (meganrogge) [inline src/vs/workbench/browser/parts/editor/accessibilityStatus.ts:28]:
```suggestion
		@IStatusbarService private readonly statusbarService: IStatusbarService
```

File: src/vs/workbench/browser/parts/editor/accessibilityStatus.ts [hunk 5]
```diff
@@ -28,7 +28,6 @@ export class AccessibilityStatus extends Disposable implements IWorkbenchContrib
 		@IStatusbarService private readonly statusbarService: IStatusbarService
 	) {
 		super();
-
 		this._register(this._accessibilityService.onDidChangeScreenReaderOptimized(() => this.onScreenReaderModeChange()));
```
Comment (meganrogge) [inline src/vs/workbench/browser/parts/editor/accessibilityStatus.ts:31]:
```suggestion

		this._register(this._accessibilityService.onDidChangeScreenReaderOptimized(() => this.onScreenReaderModeChange()));
```

File: src/vs/code/electron-main/app.ts [hunk 6]
```diff
@@ -315,6 +315,11 @@ export class CodeApplication extends Disposable {
 
 		// Dispose on shutdown
 		this.lifecycleMainService.onWillShutdown(() => this.dispose());
+		this.lifecycleMainService.onWillLoadWindow(e => {
```
Comment (meganrogge) [inline src/vs/code/electron-main/app.ts:318]: I tried moving this into the electron-sandbox `AccessibilityService`, but was having trouble bc it said `LifeCycleMainService` is not registered
Screenshot 2022-12-08 at 5 54 10 PM

**Comment by bpasero** [comment]
@meganrogge there is a way to update the configuration explicitly when doing a window reload, see here:

https://github.com/microsoft/vscode/blob/422b581e3802e30cfb780c21d9cc1a2cd0c9f0aa/src/vs/platform/windows/electron-main/windowImpl.ts#L1043

For example we update some properties here:

https://github.com/microsoft/vscode/blob/422b581e3802e30cfb780c21d9cc1a2cd0c9f0aa/src/vs/platform/windows/electron-main/windowImpl.ts#L1069-L1070

I would think we can set accessibility mode too.

**Comment by meganrogge** [comment]
Thanks for the code pointers and continued discussion. 

Based on logging, the values on `environmentService.window` do not get updated on `reload` and I don't think that's possible

It seems that consumers of `INativeWorkbenchEnvironmentService` end up using the storage service for this case instead 

For example:
https://github.com/microsoft/vscode/blob/60bbd421a19e0443852712a8c74ce056a225e59a/src/vs/workbench/services/themes/electron-sandbox/nativeHostColorSchemeService.ts#L28-L59

When I open a window, the property is known. When I changed the theme to light and reloaded, logging the value of the theme shows that it was not correct (thus the necessity of the storage service)

Screenshot 2022-12-09 at 11 02 58 AM

**Comment by bpasero** [comment]
@meganrogge theming might do something very special for other reasons, do not take it as an example. The values of the environment service will be updated based on the main side configuration from the `reload` call, can you try it?

**Comment by meganrogge** [comment]
I tried that and the value was `undefined` - perhaps because of a timing issue?

**Comment by meganrogge** [comment]
I noticed `continueOn` also uses the storage service and `policiesData` is used in `initServices` - which we can't do for the `accessibilityService`

**Comment by bpasero** [comment]
@meganrogge a simple example works for me, here are my changes:

In [`windowImpl.ts`](https://github.com/microsoft/vscode/blob/c2bb0cdb01fd7c52c62ae30da7a3f11ff6b1e8d7/src/vs/platform/windows/electron-main/windowImpl.ts#L1069):
```ts
configuration.verbose = true;
```

In constructor of [`window.ts`](https://github.com/microsoft/vscode/blob/c2bb0cdb01fd7c52c62ae30da7a3f11ff6b1e8d7/src/vs/workbench/electron-sandbox/window.ts#L128):
```ts
console.log(this.environmentService.verbose);
```

On first start I get `false` and on reload `true`.

**Comment by meganrogge** [comment]
It does appear to be working for me now

**Comment by meganrogge** [comment]
Oh I had my old changes still, hopefully it will work when I remove them as well

**Review by bpasero — CHANGES_REQUESTED** [review]
You do not need the changes in `argv.ts` if you add the property to `INativeWindowConfiguration`: https://github.com/microsoft/vscode/blob/d1f42f4857c638aa1580f0ea6d0855618f7d47ad/src/vs/platform/window/common/window.ts#L276

**Review by bpasero — CHANGES_REQUESTED** [review]
We still need to somewhere read the `accessibilitySupport` value in the workbench, no?

**Comment by meganrogge** [comment]
that already happens here

https://github.com/microsoft/vscode/blob/84b9f9386aa15a6f4af12d570648dcdf767660d3/src/vs/workbench/services/accessibility/electron-sandbox/accessibilityService.ts#L44

**Review by bpasero — APPROVED** [review]
Cool!

**Comment by meganrogge** [comment]
Thanks for the help @bpasero 👍🏼

**Pull request merged**